Fix `null` values, and add support for filtered counts

fbertsch opened the pull request.

fbertsch committed:
Update spark to 2.1.1

This is the most recent version available on EMR.

vitillo commented on src/main/scala/com/mozilla/spark/sql/hyperloglog/aggregates/HyperLogLogMerge.scala:
```diff
@@ -80,3 +80,20 @@ class HyperLogLogMerge extends UserDefinedAggregateFunction {
+ * another Boolean column and only adds the
+ * associated hll if the Boolean column is `True`.
+ */
+class FilteredHyperLogLogMerge extends HyperLogLogMerge {
```
Could you please add some tests for this new functionality?

fbertsch replied:
See https://github.com/fbertsch/spark-hyperloglog/blob/ea70f6d91b630a65166994e0d1a23770edd898d1/src/test/scala/com/mozilla/spark/sql/hyperloglog/test/HyperLogLog.scala#L77

vitillo commented on src/main/scala/com/mozilla/spark/sql/hyperloglog/aggregates/HyperLogLogMerge.scala:
```diff
@@ -80,3 +80,20 @@ class HyperLogLogMerge extends UserDefinedAggregateFunction {
 }
+
+/**
+ * This class adds the capability to take in
```
It might be worth mentioning what the use case is within the comment as it's not entirely clear otherwise.

vitillo commented on src/test/scala/com/mozilla/spark/sql/hyperloglog/test/HyperLogLog.scala:
```diff
@@ -15,31 +15,78 @@ package com.mozilla.spark.sql.hyperloglog.test
+    val rows = predata.toDS()
+      .selectExpr("os = 'windows' as allowed", "hll_create(id, 12) as hll")
+      .groupBy()
+      .agg(expr("hll_cardinality(hll_filtered_merge(hll, allowed)) as count"))
```
Hm, I just noticed these tests; they should be part of the last commit.

fbertsch committed:
Handle null and empty values

fbertsch committed:
Add support for filtered counts

Values are only added when an associated Boolean column is `True`.
This makes it easy to support % of client queries, since the
query is only
```
cardinality(filtered_client_hll) / cardinality(client_hll)
```
rather than complicated joins if this were included as a
dimension in the group by.
